File: Dynamic_Programming/C/fibnacci.c
/*
Author: Kumar Shanu
Program: nth fibnacci number
*/

#include <stdio.h>
#include <string.h>

// recursive function for nth fibnacci number
// Time Complexity: O(2^n)
int fib(int n)
{
    if (n <= 1)
    {
        return n;
    }
    else
    {
        return fib(n - 1) + fib(n - 2);
    }
}

// fibnacci recursive function with memorization
int memo[50];
int fib(int n)
{
    if (n <= 1)
        return n;

    int first, second;

    if (memo[n - 1] != -1)
        first = memo[n - 1];
    else
        first = fib(n - 1);

    if (memo[n - 2] != -1)
        second = memo[n - 2];
    else
        second = fib(n - 2);

    return memo[n] = first + second;
}

// iterative function using dynamic programming
// Time Complexity: O(n)
// Space: O(n)
int fib(int n)
{
    int dp[n + 1];
    dp[0] = 0;
    dp[1] = 1;
    for (int i = 2; i <= n; i++)
    {
        dp[i] = dp[i - 1] + dp[i - 2];
    }
    return dp[n];
}

int main()
{
    int n;
    memset(memo, -1, sizeof(memo));

    printf("Enter the value of n: ");
    scanf("%d", &n);
    printf("\nnth fibnacci number %d\n", fib(n));

    return 0;
}

File: Dynamic_Programming/python/fibnacci.py
"""
KUMAR SHANU
Program: find nth fibnacci number
"""

# recursive function for nth fibnacci number


def fib(n):
    return (1, 0)[n < 1] if n < 2 else fib(n-1) + fib(n-2)


# recursive function using memorization
memo = [-1]*100


def fib(n):
    if n <= 1:
        return n
    if memo[n-1] != -1:
        first = memo[n-1]
    else:
        first = fib(n-1)

    if memo[n-2] != -1:
        second = memo[n-2]
    else:
        second = fib(n-2)

    memo[n] = first + second
    return memo[n]


# iterative function using dynamic programming

def fib(n):
    dp = [-1]*(n+1)
    dp[0], dp[1] = 0, 1

    for i in range(2, n+1):
        dp[i] = dp[i-1] + dp[i-2]

    return dp[n]


if __name__ == "__main__":
    n = int(input("Enter n: "))
    print("nth fibnacci number is: ", fib(n))
